compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs

```diff
@@ -3018,12 +3018,16 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
                                 [] => span_bug!(ty.span, "trait object with no traits: {ty:?}"),
                             };
                             let needs_parens = traits.len() != 1;
-                            err.span_suggestion_verbose(
-                                span,
-                                "you can use `impl Trait` as the argument type",
-                                "impl ",
-                                Applicability::MaybeIncorrect,
-                            );
+                            if let Some(hir_id) = hir_id
+                                && self.tcx.parent_hir_node(hir_id).ident().is_some()
```

Comment:
`.ident().is_some()` includes (and does not include) a lot of things. it's better for clarity and exhaustiveness to do something like this:

`!matches!(node, Node::Expr(Expr { kind: ExprKind::Closure(_), .. }))`

or (better) to enumerate where this suggestion *is* valid:

`matches!(node, /* everywhere impl Trait argument is valid */)`

Reply:
Indeed ! I have tried to use the second (better) approach and the right type